src/background.c

```diff
@@ -409,6 +474,372 @@ background_class_init (BackgroundClass *klass)
+
+  xdp_impl_background_call_notify_background (background_impl,
+                                              handle,
+                                              ddata->app_id,
+                                              name,
+                                              NULL,
+                                              notify_background_done,
+                                              ddata);
+}
+
+static void
+check_background_apps (void)
+{
+  g_autoptr(GVariant) perms = NULL;
+  g_autoptr(GHashTable) app_states = NULL;
+  g_autoptr(GPtrArray) apps = NULL;
+  int i;
+
+  app_states = get_app_states ();
+  if (app_states == NULL)
+    return;
+
+  g_debug ("Checking background permissions");
+
+  perms = get_all_permissions ();
+  apps = flatpak_instance_get_all ();
+
+  remove_outdated_applications (apps);
+
+  for (i = 0; i < apps->len; i++)
+    {
+      FlatpakInstance *instance = g_ptr_array_index (apps, i);
+      const char *app_id = flatpak_instance_get_app (instance);
+      Permission permission;
+      AppState state;
+      const char *state_names[] = { "background", "running", "active" };
+      g_autofree char *handle = NULL;
+
+      if (!flatpak_instance_is_running (instance))
+        continue;
+
+      state = get_one_app_state (app_id, app_states);
+      g_debug ("App %s is %s", app_id, state_names[state]);
+
+      if (state != BACKGROUND)
+        {
+          remove_background_app (app_id);
+          continue;
+        }
+
+      /* If the app is not in the list of background apps
+       * yet, add it, but don't notify yet - this gives
+       * apps some leeway to get their window app. If it
+       * is still in the background next time around,
+       * we'll proceed to the next step.
+       */
+      if (!lookup_background_app (app_id, &handle))
+        {
+          add_background_app (app_id, NULL);
+          continue;
+        }
+
+      if (handle)
+        continue; /* already notified */
+
+      permission = get_one_permission (app_id, perms);
+      if (permission == NO)
+        {
+          pid_t pid = flatpak_instance_get_child_pid (instance);
+          g_debug ("Killing app %s (child pid %u)", app_id, pid);
+          kill (pid, SIGKILL);
+        }
+      else if (permission == ASK || permission == UNSET)
+        {
+          send_notification (instance, permission);
+        }
+    }
+}
+
+static void
+background_monitor (GTask *task,
+                    gpointer source_object,
+                    gpointer task_data,
+                    GCancellable *cancellable)
+{
+  while (1)
+    {
+      check_background_apps ();
+      sleep (60);
+    }
+}
+
+static void
+start_background_monitor (void)
+{
+  g_autoptr(GTask) task = NULL;
+
+  applications = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+  g_debug ("Starting background app monitor");
+
+  task = g_task_new (NULL, NULL, NULL, NULL);
+  g_task_run_in_thread (task, background_monitor);
```

Comment:
I don't think you should use a GTask for this. That will constantly use one of the threads in the GTask threadpool. Just spawn a regular thread?